File: engine/src/engine/core/temporal.ts
import { ManseryeokRangeError } from './errors';

const MILLISECONDS_PER_MINUTE = 60_000;
const MILLISECONDS_PER_DAY = 86_400_000;
const KST_OFFSET_MINUTES = 9 * 60;

export interface DateTimeParts {
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
  second: number;
}

function assertInteger(value: number, label: string): void {
  if (!Number.isInteger(value)) {
    throw new ManseryeokRangeError(`${label} must be an integer`, { label, value });
  }
}

function timestampFromUtcParts(dateTime: DateTimeParts): number {
  assertInteger(dateTime.year, 'year');
  assertInteger(dateTime.month, 'month');
  assertInteger(dateTime.day, 'day');
  assertInteger(dateTime.hour, 'hour');
  assertInteger(dateTime.minute, 'minute');
  assertInteger(dateTime.second, 'second');

  const date = new Date(Date.UTC(
    dateTime.year,
    dateTime.month - 1,
    dateTime.day,
    dateTime.hour,
    dateTime.minute,
    dateTime.second,
  ));
  date.setUTCFullYear(dateTime.year);
  const timestamp = date.getTime();

  if (!Number.isFinite(timestamp)) {
    throw new ManseryeokRangeError('Date-time is outside the supported JavaScript UTC range', { dateTime });
  }

  const normalized = partsFromUtcTimestamp(timestamp);
  if (compareDateTime(normalized, dateTime) !== 0) {
    throw new ManseryeokRangeError('Date-time contains invalid UTC calendar fields', { dateTime });
  }

  return timestamp;
}

function partsFromUtcTimestamp(timestamp: number): DateTimeParts {
  const date = new Date(timestamp);
  return {
    year: date.getUTCFullYear(),
    month: date.getUTCMonth() + 1,
    day: date.getUTCDate(),
    hour: date.getUTCHours(),
    minute: date.getUTCMinutes(),
    second: date.getUTCSeconds(),
  };
}

function pad(value: number, width: number): string {
  return String(value).padStart(width, '0');
}

export function shiftDateTimeUtc(dateTime: DateTimeParts, minutes: number): DateTimeParts {
  assertInteger(minutes, 'minutes');
  return partsFromUtcTimestamp(timestampFromUtcParts(dateTime) + minutes * MILLISECONDS_PER_MINUTE);
}

export function toOffsetTimestamp(dateTime: DateTimeParts, offsetMinutes: number): number {
  assertInteger(offsetMinutes, 'offsetMinutes');
  return timestampFromUtcParts(dateTime) - offsetMinutes * MILLISECONDS_PER_MINUTE;
}

export function toKstTimestamp(dateTime: DateTimeParts): number {
  return toOffsetTimestamp(dateTime, KST_OFFSET_MINUTES);
}

export function dayOfYearUtc(dateTime: Pick<DateTimeParts, 'year' | 'month' | 'day'>): number {
  const current = timestampFromUtcParts({
    year: dateTime.year,
    month: dateTime.month,
    day: dateTime.day,
    hour: 0,
    minute: 0,
    second: 0,
  });
  const firstDay = timestampFromUtcParts({
    year: dateTime.year,
    month: 1,
    day: 1,
    hour: 0,
    minute: 0,
    second: 0,
  });

  return Math.floor((current - firstDay) / MILLISECONDS_PER_DAY) + 1;
}

export function compareDateTime(left: DateTimeParts, right: DateTimeParts): number {
  const fields = ['year', 'month', 'day', 'hour', 'minute', 'second'] as const;
  for (const field of fields) {
    if (left[field] < right[field]) {
      return -1;
    }
    if (left[field] > right[field]) {
      return 1;
    }
  }

  return 0;
}

export function formatDateKey(dateTime: Pick<DateTimeParts, 'year' | 'month' | 'day'>): string {
  return `${pad(dateTime.year, 4)}-${pad(dateTime.month, 2)}-${pad(dateTime.day, 2)}`;
}

export function diffMinutes(from: DateTimeParts, to: DateTimeParts): number {
  return Math.trunc((timestampFromUtcParts(to) - timestampFromUtcParts(from)) / MILLISECONDS_PER_MINUTE);
}

File: realapp/lib/engine/manseryeok/temporal.dart
/// Port of engine/src/engine/core/temporal.ts.
library;

import 'errors.dart';

const int _msPerMinute = 60000;
const int kstOffsetMinutes = 9 * 60;

class DateTimeParts {
  const DateTimeParts({
    required this.year,
    required this.month,
    required this.day,
    this.hour = 0,
    this.minute = 0,
    this.second = 0,
  });

  final int year;
  final int month;
  final int day;
  final int hour;
  final int minute;
  final int second;

  DateTimeParts copyWith({int? hour, int? minute, int? second}) =>
      DateTimeParts(
        year: year,
        month: month,
        day: day,
        hour: hour ?? this.hour,
        minute: minute ?? this.minute,
        second: second ?? this.second,
      );

  @override
  String toString() => '$year-$month-$day $hour:$minute:$second';
}

int timestampFromUtcParts(DateTimeParts p) {
  final date = DateTime.utc(p.year, p.month, p.day, p.hour, p.minute, p.second);
  final normalized = partsFromUtcTimestamp(date.millisecondsSinceEpoch);
  if (compareDateTime(normalized, p) != 0) {
    throw const ManseryeokException(
      ManseryeokErrorCode.rangeError,
      'Date-time contains invalid UTC calendar fields',
    );
  }
  return date.millisecondsSinceEpoch;
}

DateTimeParts partsFromUtcTimestamp(int timestamp) {
  final d = DateTime.fromMillisecondsSinceEpoch(timestamp, isUtc: true);
  return DateTimeParts(
    year: d.year,
    month: d.month,
    day: d.day,
    hour: d.hour,
    minute: d.minute,
    second: d.second,
  );
}

DateTimeParts shiftDateTimeUtc(DateTimeParts dateTime, int minutes) {
  return partsFromUtcTimestamp(
    timestampFromUtcParts(dateTime) + minutes * _msPerMinute,
  );
}

int toKstTimestamp(DateTimeParts dateTime) {
  return timestampFromUtcParts(dateTime) - kstOffsetMinutes * _msPerMinute;
}

int compareDateTime(DateTimeParts left, DateTimeParts right) {
  final pairs = [
    (left.year, right.year),
    (left.month, right.month),
    (left.day, right.day),
    (left.hour, right.hour),
    (left.minute, right.minute),
    (left.second, right.second),
  ];
  for (final (l, r) in pairs) {
    if (l < r) return -1;
    if (l > r) return 1;
  }
  return 0;
}

String formatDateKey(DateTimeParts dateTime) {
  final y = dateTime.year.toString().padLeft(4, '0');
  final m = dateTime.month.toString().padLeft(2, '0');
  final d = dateTime.day.toString().padLeft(2, '0');
  return '$y-$m-$d';
}
